File: algorithms/mergesort.py
from collections import deque

def mergesort(array, low, high):
    if low < high:
        middle = (low + high) / 2
        mergesort(array, low, middle)
        mergesort(array, middle+1, high)
        merge(array, low, middle, high)

def merge(array, low, middle, high):

    bufferLow = deque(array[low:middle+1])
    bufferHigh = deque(array[middle+1:high+1])

    i = 0

    while bufferLow and bufferHigh:
        if bufferLow[0] < bufferHigh[0]:
            array[i] = bufferLow.popleft()
        else:
            array[i] = bufferHigh.popleft()

        i += 1

    while bufferLow:
        array[i] = bufferLow.popleft()
        i += 1

    while bufferHigh:
        array[i] = bufferHigh.popleft()
        i+= 1

def main():
    array = [43, 251, 763, 1324, 5, 27, 721, 62, 346]
    mergesort(array, 0, len(array) - 1)
    print array

main()

File: algorithms/mergesort.c
#include <stdio.h>

void merge(int array[], int low, int middle, int high) {

    int buffer[high - low + 1];

    int i = 0;
    int lowPos = low;
    int highPos = middle + 1;

    while (lowPos <= middle && highPos <= high) {
        if (array[lowPos] < array[highPos]) {
            buffer[i++] = array[lowPos++];
        }
        else {
            buffer[i++] = array[highPos++];
        }
    }

    while (lowPos <= middle) buffer[i++] = array[lowPos++];
    while (highPos <= high) buffer[i++] = array[highPos++];

    int index = low;
    while (index <= high) {
        array[index] = buffer[index - low];
        ++index;
    }
}


void mergesort(int array[], int low, int high) {
    int middle;

    if (low >= high) return;

    middle = (low + high) / 2;

    mergesort(array, low, middle);
    mergesort(array, middle + 1, high);
    merge(array, low, middle, high);
}


int main() {
    int array[10] = { 123, 542, 12, 76, 32, 211, 356, 864, 1, 721 };
    
    mergesort(array, 0, 9);

    int i = 0;
    while (i < 10) {
        printf("%d ", array[i++]);
    }
    printf("\n");
}



